File: python/y2022/d03/aoc_2022_03.py
import sys
from common import timer, Puzzle


def prio(char: str):
    if char.isupper():
        return ord(char) - ord('A') + 27
    return ord(char) - ord('a') + 1


class Day3(Puzzle, year=2022, day=3):
    @staticmethod
    def clean_input(data):
        rucksacks = list()
        for r in data:
            row = r.strip()
            rucksacks.append((row[0:int(len(row)/2)], row[int(len(row)/2):]))
        return rucksacks

    @timer(part=1)
    def solve_part_1(self):
        """Solution for part 1"""
        prio_sum = 0
        for item1, item2 in self.data:
            set_of_items = set(item1).intersection(set(item2))
            for common_item in set_of_items:
                prio_sum += prio(common_item)
        return prio_sum

    @timer(part=2)
    def solve_part_2(self):
        """Solution for part 2"""
        groups = list()
        group = list()

        for group_index, pack in enumerate(self.data):
            group.append(''.join(pack))
            if ((group_index + 1) % 3) == 0 and group_index > 0:
                groups.append(group)
                group = list()

        prio_sum = 0
        for group in groups:
            group_set = set(group[0])
            for member in group[1:]:
                group_set = group_set.intersection(set(member))
            prio_sum += prio(group_set.pop())

        return prio_sum


if __name__ == "__main__":
    filename = "input.txt"
    if len(sys.argv) > 1:
        filename = sys.argv[1]
    Day3(filename=filename).solve_all()

File: C++/2022/03/aoc_2022_03.cpp
#include "aoc_io.hpp"
#include "aoc_timer.hpp"
#include <algorithm>
#include <iterator>
#include <set>
#include <string>
#include <vector>

namespace aoc_2022_03 {

static int prio(char p) {
    if (std::isupper(p)) {
        return p - 'A' + 27;
    }
    return p - 'a' + 1;
}

std::vector<std::pair<std::string, std::string>> clean_input(const std::vector<std::string> &input_data) {
    std::vector<std::pair<std::string, std::string>> rucksacks;
    rucksacks.reserve(input_data.size());
    for (auto &row : input_data) {
        std::pair<std::string, std::string> packs;
        packs.first = row.substr(0, row.length() / 2);
        packs.second = row.substr(row.length() / 2, row.length() / 2);
        rucksacks.push_back(packs);
    }
    return rucksacks;
}

int solve_part_1(const std::vector<std::pair<std::string, std::string>> &data) {
    std::int32_t prio_sum = 0;

    for (auto &pack : data) {
        std::set<char> first(pack.first.begin(), pack.first.end());
        std::set<char> second(pack.second.begin(), pack.second.end());

        std::vector<char> common;
        std::ranges::set_intersection(first, second, std::back_inserter(common));
        prio_sum += prio(common[0]);
    }

    return prio_sum;
}

void *solve_all(const std::vector<std::string> &data) {
    std::vector<std::pair<std::string, std::string>> input;

    input = clean_input(data);

    if (data.size() > 0) {
        aoc::timer(1, solve_part_1, input, 1);
        // aoc::timer(2, solve_part_2, data, 1);
    }

    return NULL;
}

} // namespace aoc_2022_03

int main(int argc, char **argv) {
    std::vector<std::string> data;

    char sourcefile[20];
    int  year = 2022;
    int  day = 3;

    if (argc > 1) {
        if (std::string(argv[1]) == "--test") {
            data = aoc::io::get_input_list<std::string>("test_input.txt", year, day);
        } else {
            data = aoc::io::get_input_list<std::string>(argv[1], year, day);
        }
    } else {
        data = aoc::io::get_input_list<std::string>("input.txt", year, day);
    }

    aoc::io::header(year, day);
    aoc::timer(0, aoc_2022_03::solve_all, data, 0);

    return 0;
}
